subprojects/launcher/src/main/java/org/gradle/launcher/cli/CommandLineActionFactory.java

```diff
@@ -110,6 +117,69 @@ private static void showUsage(PrintStream out, CommandLineParser parser) {
         out.println();
     }
 
+    private static class WelcomeMessageAction implements Action<PrintStream> {
+        private final BuildLayoutParameters buildLayoutParameters;
+
+        public WelcomeMessageAction(BuildLayoutParameters buildLayoutParameters) {
+            this.buildLayoutParameters = buildLayoutParameters;
+        }
+
+        @Override
+        public void execute(PrintStream out) {
+            GradleVersion currentVersion = GradleVersion.current();
+            File gradleUserHomeDir = buildLayoutParameters.getGradleUserHomeDir();
+            File notificationsDir = new File(gradleUserHomeDir, "notifications/" + currentVersion.getVersion());
+            File markerFile = new File(notificationsDir, "release-features.rendered");
+
+            if (!markerFile.exists()) {
+                out.println();
+                out.print("Welcome to Gradle " + currentVersion.getVersion() + "!");
+
+                String featureList = readReleaseFeatures();
+
+                if (featureList != null) {
+                    out.println();
+                    out.println();
+                    out.print("Here is what's new:");
+                    out.println();
+                    out.print(featureList);
+                }
+
+                if (!currentVersion.isSnapshot()) {
+                    out.println();
+                    out.print("For more details see https://gradle.org/releases/#" + currentVersion.getVersion());
```

Comment:
Any particular reason we're using `print()` instead of `println()` here?

Reply:
I've stuck to the way we rendered message in other places of the class e.g. here: https://github.com/gradle/gradle/blob/master/subprojects/launcher/src/main/java/org/gradle/launcher/cli/CommandLineActionFactory.java#L103.

Reply:
This is a nitpick. I'm fine leaving as-is

Reply:
Tried to switch to the Gradle logger but some other tests were failing. Didn't proceed on that route.